[UnifiedPDF] Select element for choice annotations will sometimes flicker

sammygill opened the pull request:
#### b7d0e2efd37914f2c13e5a148dc3213386ac3bc6
<pre>
[UnifiedPDF] Select element for choice annotations will sometimes flicker
<a href="https://bugs.webkit.org/show_bug.cgi?id=274634">https://bugs.webkit.org/show_bug.cgi?id=274634</a>
<a href="https://rdar.apple.com/128406464">rdar://128406464</a>

Reviewed by Abrar Rahman Protyasha.

The choice annotation has a HTML select element backing it for user
interaction. Sometimes clicking on the choice annotation while the
corresponding HTML element is already active will result in the
element disappearing and reappearing resulting in a flicker. It appears
this is due to the logic in WebPageMac::performImmediateActionHitTestAtLocation
taking focus away from the HTML element and giving it to the plugin
document.

To fix this let the plugin note that the the hit test is over an already
active annotation via WebHitTestResultData. If this is the case then
we will not focus the plugin document and keep the focus on the
HTML annotation.

* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::textForImmediateActionHitTestAtPoint):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):

Canonical link: <a href="https://commits.webkit.org/279381@main">https://commits.webkit.org/279381@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05e2c8f9587be8e025cee9ea159ce93d94b684e7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53051 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32388 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5538 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56330 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3774 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39248 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3500 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43024 "Found 1 new test failure: http/tests/scroll-to-text-fragment/no-scroll-after-stylesheet-load.html (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2438 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55149 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30143 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45796 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24152 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27156 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1933 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49008 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3275 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57925 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3233 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50420 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29412 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46016 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49720 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11623 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29166 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
```diff
@@ -930,7 +930,14 @@ static void drawPDFPage(PDFDocument *pdfDocument, CFIndex pageIndex, CGContextRe
                 // FIXME (144030): Focus does not seem to get set to the PDF when invoking the menu.
-                if (RefPtr pluginDocument = dynamicDowncast<PluginDocument>(element->document()))
+                RefPtr pluginDocument = dynamicDowncast<PluginDocument>(element->document());
+                auto shouldFocusPluginDocument = [&pluginDocument, &immediateActionResult] {
+                    if (!pluginDocument)
+                        return false;
+                    return !immediateActionResult.isActivePDFAnnotation;
+                }();
```
Optional, if you want to be a little concise. I don't think it harms readability too much:

```suggestion
                bool shouldFocusPluginDocument = pluginDocument ? !immediateActionResult.isActivePDFAnnotation : false;
```

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -3700,10 +3700,15 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
             continue;
 
-        if (!FloatRect { [annotation bounds] }.contains(pagePoint))
+        if (m_activeAnnotation && m_activeAnnotation->annotation() == annotation)
+            data.isActivePDFAnnotation = true;
```
Since we don't initialize `isActivePDFAnnotation` to `false` in the structure declaration, I suppose we should always populate it with a valid boolean, lest the garbage value it's initialized with produces unexpected side effects.

```suggestion
        data.isActivePDFAnnotation = m_activeAnnotation && m_activeAnnotation->annotation() == annotation;
```

aprotyas replied:
Actually, we want to `data.isActivePDFAnnotation = false` as soon as we enter `UnifiedPDFPlugin::textForImmediateActionHitTestAtPoint()`

sammygill replied:
Good point! It looks like the constructors for this struct are handling the initialization of its members so I did the same

sammygill committed:
[UnifiedPDF] Select element for choice annotations will sometimes flicker
https://bugs.webkit.org/show_bug.cgi?id=274634
rdar://128406464

Reviewed by Abrar Rahman Protyasha.

The choice annotation has a HTML select element backing it for user
interaction. Sometimes clicking on the choice annotation while the
corresponding HTML element is already active will result in the
element disappearing and reappearing resulting in a flicker. It appears
this is due to the logic in WebPageMac::performImmediateActionHitTestAtLocation
taking focus away from the HTML element and giving it to the plugin
document.

To fix this let the plugin note that the the hit test is over an already
active annotation via WebHitTestResultData. If this is the case then
we will not focus the plugin document and keep the focus on the
HTML annotation.

* Source/WebKit/Shared/WebHitTestResultData.cpp:
(WebKit::WebHitTestResultData::WebHitTestResultData):
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/Shared/WebHitTestResultData.serialization.in:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::textForImmediateActionHitTestAtPoint):
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):

Canonical link: https://commits.webkit.org/279381@main
